### ISSUE TITLE
Wasm MVVM ViewModel (ViewModelLocator) DI constructor with parameter freeze

### ISSUE THREAD

**Issue by Dimigergo** [unoplatform/uno#5386]
## Current behavior

In the MLR.WiPlan.Cilent.Shared project we have a CMainViewModel.cs in ViewModels.
In the line 73 there is the constructor:
`public CMainViewModel(Shared.Proto.UserManagement.CUserManagement.CUserManagementClient umGrpcClient)`

The ViewModelLocator creates the instance:
```
    public class CViewModelLocator
    {
        public CViewModelLocator()
        {
        }

        public CMainViewModel MainViewModel
        {
            get { return (CMainViewModel)ActivatorUtilities.GetServiceOrCreateInstance(((App)App.Current).Container, typeof(CMainViewModel)); }
        }
    }
```

In App.xaml.cs we initialized the dependency injection classes:
`services.AddGrpcClient<Shared.Proto.UserManagement.CUserManagement.CUserManagementClient>(`

Everything build perfectly, and the wasm project freeze with this constructor. In the WPF application it is perfectly works.

## How to reproduce it (as minimally and precisely as possible)

[wasm_constructor_di.zip](https://github.com/unoplatform/uno/files/6081888/wasm_constructor_di.zip)

## Environment

Microsoft Visual Studio Professional 2019 Preview
Version 16.10.0 Preview 1.0
VisualStudio.16.Preview/16.10.0-pre.1.0+31025.218
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Professional

Visual C++ 2019   00435-60000-00000-AA607
Microsoft Visual C++ 2019

.NET Core Debugging with WSL 2   1.0
.NET Core Debugging with WSL 2

ADL Tools Service Provider   1.0
This package contains services used by Data Lake tools

ASA Service Provider   1.0

ASP.NET and Web Tools 2019   16.10.127.22515
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2111611+54ccb705f82f77549dbc770a598e1c42b4c166ce
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2019   16.10.127.22515
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.10.127.22515
Azure App Service Tools v3.0.0

Azure Data Lake Node   1.0
This package contains the Data Lake integration nodes for Server Explorer.

Azure Data Lake Tools for Visual Studio   2.6.3000.0
Microsoft Azure Data Lake Tools for Visual Studio

Azure Functions and Web Jobs Tools   16.10.127.22515
Azure Functions and Web Jobs Tools

Azure Stream Analytics Tools for Visual Studio   2.6.3000.0
Microsoft Azure Stream Analytics Tools for Visual Studio

C# Tools   3.10.0-1.21125.6+93b8af2b1f2c9d6318f0fceddc04bb736684bf66
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Fabric.DiagnosticEvents   1.0
Fabric Diagnostic Events

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

License Header Manager   3.0.3
An extension for automatically inserting license information file headers.

Microsoft Azure HDInsight Azure Node   2.6.3000.0
HDInsight Node under Azure Node

Microsoft Azure Hive Query Language Service   2.6.3000.0
Language service for Hive query

Microsoft Azure Service Fabric Tools for Visual Studio   16.0
Microsoft Azure Service Fabric Tools for Visual Studio

Microsoft Azure Stream Analytics Language Service   2.6.3000.0
Language service for Azure Stream Analytics

Microsoft Azure Stream Analytics Node   1.0
Azure Stream Analytics Node under Azure Node

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.40218.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   16.9.7 (df23ba6)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.10.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.62102.01130
Microsoft SQL Server Data Tools

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript Tools   16.0.30217.2001
TypeScript Tools for Microsoft Visual Studio

UnoPlatformPackage Extension   1.0
UnoPlatformPackage Visual Studio Extension Detailed Info

Visual Basic Tools   3.10.0-1.21125.6+93b8af2b1f2c9d6318f0fceddc04bb736684bf66
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   16.10.0-beta.21117.1+b8515e673596623e15e74a4e6fe537994b21b6e2
Microsoft Visual F# Tools

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Kubernetes   1.0
Visual Studio Tools for Kubernetes

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Foo   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.10.000.21 (main@240d4e4)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.10.0.0 (remotes/origin/8affbf0abc24122abd5517b2b52f1073cb52bd7b@8affbf0ab)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.9.75 (e29e63c)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   11.2.0.21 (d16-9/93eab59)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: 5e9cb6d
    Java.Interop: xamarin/java.interop/d16-9@d6d86b2
    ProGuard: Guardsquare/proguard/v7.0.1@912d149
    SQLite: xamarin/sqlite/3.34.1@daff8f4
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-9@9d8924d

Xamarin.iOS and Xamarin.Mac SDK   14.14.2.5 (3836759d4)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

Affected platform(s):

- [ ] iOS
- [ ] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

IDE:

- [ ] Visual Studio 2017 (version: )
- [x] Visual Studio 2019 (version: )
- [ ] Visual Studio for Mac (version: )
- [ ] Rider Windows (version: )
- [ ] Rider macOS (version: )
- [ ] Visual Studio Code (version: )

**Comment by jeromelaban** [comment]
Thanks for the report. This may very well be an exception related to the linker. Can you try referring to [this section](https://platform.uno/docs/articles/get-started-wizard.html#runtime-error-no-parameterless-constructor-defined-for-xxxx) about the linker configuration ?

**Comment by Dimigergo** [comment]
Thanks for the suggestion, it wors perfectly!

If I switch off the linker trimming with
`<WasmShellILLinkerEnabled>false</WasmShellILLinkerEnabled>`
the code works perfect.

After that I enabled it, and caught the exception, which has a line:
```System.InvalidOperationException: A suitable constructor for type 'MLR.WiPlan.Shared.Proto.UserManagement.CUserManagement+CUserManagementClient' could not be located. Ensure the type is concrete and services are registered for all parameters of a public constructor.```

This CUserManagementClient class as can see in the demo repro is in the MLR.WiPlan.Shared.Proto assembly, not in the MLR.WiPlan.Cilent.Shared assembly.
So I added the line:
`<assembly fullname="MLR.WiPlan.Shared.Proto" />`
to my LinkerConfig.xml, and everything is works perfectly.
